src/packages/text_search.gleam

```diff
@@ -46,45 +48,113 @@ pub fn update(
+        Ok(Packages(packages: rank_results(packages, phrase)))
+
+      // If there's no results matching the given query we try and guess what
+      // the user might have wanted to type.
+      [] -> {
+        use words <- result.try(
+          ethos.keys(index.table)
+          |> result.replace_error(error.EtsTableError),
+        )
+        case did_you_mean(given: phrase, one_of: words) {
+          Ok(suggestion) -> Ok(DidYouMean(suggestion:))
+          Error(_) -> Ok(Packages(packages: []))
+        }
+      }
+    }
+  })
+}
+
+fn did_you_mean(
+  given phrase: String,
+  one_of words: List(String),
+) -> Result(String, Nil) {
+  // We want to limit the maximum edit distance. Otherwise we could end up
+  // suggesting fixes that are not related at all to the original query.
+  let phrase_length = string.length(phrase)
+  let limit = int.max(1, phrase_length / 3)
+
+  list.filter_map(words, fn(word) {
+    let word_length = string.length(word)
+    let minimum_distance = int.absolute_value(word_length - phrase_length)
+
+    // If the minimum distance is greater than the allowed limit then we don't
+    // even waste any time computing the edit distance of the two strings!
+    use <- bool.guard(when: minimum_distance > limit, return: Error(Nil))
+    let distance = edit_distance.levenshtein(phrase, word)
+    case distance > limit {
+      True -> Error(Nil)
+      False -> Ok(#(word, distance))
+    }
+  })
+  // We only pick the word with the smallest possible edit distance that's below
+  // the given threshold
+  |> list.sort(fn(one, other) { int.compare(one.1, other.1) })
+  |> list.first
+  |> result.map(fn(suggestion) { suggestion.0 })
+}
+
+/// Given a list with packages matching the searched phrase we assign them a
+/// ranking and return them in a list with no duplicates sorted from best match
+/// to worst match.
+///
+fn rank_results(packages: List(String), phrase: String) -> List(String) {
+  let ranking =
+    // Each package starts with a score that is the number of times it appears
+    // in the search results coming from `ethos`.
+    count_occurrences(packages)
     |> dict.to_list
     |> list.map(fn(pair) {
-      case pair.0 {
-        // Rank up proritised packages
+      // We then tweak the ranking to prioritise some official Gleam packages.
+      let #(package_name, score) = pair
+      case package_name {
         "gleam_stdlib"
         | "gleam_javascript"
         | "gleam_erlang"
         | "gleam_otp"
         | "gleam_json"
-        | "gleam_time" -> #(pair.0, pair.1 + 10)
+        | "gleam_time" -> #(package_name, score + 10)
         _ -> pair
       }
     })
-    |> list.sort(fn(a, b) {
-      case a, b {
-        // Exact matches come first
-        #(name, _), _ if name == phrase -> order.Lt
-        _, #(name, _) if name == phrase -> order.Gt
-        // Otherwise compare the score
-        _, _ -> int.compare(b.1, a.1)
-      }
+
+  list.sort(ranking, fn(one, other) {
+    case one, other {
```

Comment:
Revert these names please, the two new ones are no more descriptive but they're less visually distinct from each other, and less visually distinct from `order`, so the code has become harder to scan.